### PR TITLE
feat: updating source

### PR DESCRIPTION
**Pull request by tpluscode**
@martinmaillard note that this PR has a technical commit 52bcbe6, which you might review in separation. it slightly improves the code around publishing and handling of domain events

otherwise the bulk is in commit 3aaedbe, where I added the `PUT /source` endpoint which updates the name, delimiter and quote char of a `Source`.

### CODE REVIEW

File: api/lib/handlers/update-source.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+csvSourceEvents.on.NameChanged(function updateName (ev) {
+  return deleteInsert(`<${ev.id}> schema:name ?current .`)
+    .insert(`<${ev.id}> schema:name "${ev.data.newName}"`)
+    .prefixes({ schema })
+    .execute(getClient())
+})
+
+csvSourceEvents.on.QuoteChanged(function updateQuote (ev) {
+  return deleteInsert(`<${ev.id}> dataCube:csvQuote ?current .`)
+    .insert(`<${ev.id}> dataCube:csvQuote ${toString(literal(ev.data.newQuote))}`)
```
Comment (tpluscode) [inline api/lib/handlers/update-source.ts]: A little better that the manual escape which I did in the other PR.

Still this is not super pretty. I might push for the `@zazuko/rdf-string` package soon

**Comment by martinmaillard** [comment]
From the code, I get that you can send a file with a different name? What happens to the existing attributes/table identifiers if the new file doesn't have the same columns?

**Comment by tpluscode** [comment]
No, not a file. Just the meta

**Comment by martinmaillard** [comment]
What do you mean? Does this endpoint expect something else than a file? If yes, that's going to be weird for users, no?

If they are able to change the CSV options for a source, they should probably also be able to set them manually when uploading the source. And to visualize them somewhere.

**Comment by tpluscode** [comment]
The request is only to update the RDF representation of the Source itself

```http
POST /Source/foo-csv
Content-Type: text/turtle

</Source/foo-csv>
  schema:name "foo-bar.csv";
  dataCube:csvDelimiter ";" ;
  dataCube:csvQuote "'" .
```

To re-upload the actual file will be much more complicated because it affects whole bunch of other resources. This is only to tweak how the file will be handled but not its expected contents

**Comment by martinmaillard** [comment]
When retrieving the source itself, is there any way to provide the default `csvDelimiter` and `csvQuote` if none are defined?

**Comment by tpluscode** [comment]
Of course, possible, but why do you need it?

**Comment by martinmaillard** [comment]
Because from a user's perspective, the defaults *are* the delimiter/quote when none was defined. It doesn't make sense for them to have empty fields.

But maybe it's just an issue with sources that were created in a previous version? In this case, we don't need to care about that too much I guess

**Comment by tpluscode** [comment]
Yea, I would think so. Any new source will have values discovered by the sniffer library.

And otherwise, empty values indicate that something is wrong so maybe it's good to keep it obvious.

**Comment by martinmaillard** [comment]
Alright, that makes sense

**Comment by martinmaillard** [comment]
I just tested this locally and got a 500. The following error was raised in the API:
```
2020-02-12T10:24:06.055Z dataCube:request:error Request failed: TypeError: Cannot destructure property `name` of 'undefined' or 'null'. at .../data-cube-curation/api/lib/domain/source/updateSource.ts:15:9 at .../data-cube-curation/api/node_modules/@tpluscode/fun-ddr/lib/mutation.js:8:16 at .../data-cube-curation/api/node_modules/@tpluscode/fun-ddr/lib/AggregateRootImpl.js:42:46 at runMicrotasks (<anonymous>) at processTicksAndRejections (internal/process/task_queues.js:93:5) at AggregateRootImpl.commit (.../data-cube-curation/api/node_modules/@tpluscode/fun-ddr/lib/AggregateRootImpl.js:85:9) at .../data-cube-curation/api/lib/data-cube/project-source/update.ts:20:3
```

**Comment by martinmaillard** [comment]
Now I get a 400 with "Name cannot be empty"

**Comment by tpluscode** [comment]
Good catch with `schema:name` although that was unrelated.

Here's an example body which definitely works:

```turtle
@prefix schema: <http://schema.org/> .
@prefix dataCube: <https://rdf-cube-curation.described.at/> .

<> schema:name "Changed name" ;
   dataCube:csvDelimiter "," ;
   dataCube:csvQuote "\"" .
```

**Comment by martinmaillard** [comment]
I'm still getting the same 400. Will try with a fresh store.

**Comment by martinmaillard** [comment]
Nope, doesn't solve it.

I console.logged the `command` object that is being destructured in `updateSource.ts` and it looks like:
```
UpdateModel {
  __initialized: true,
  _selfGraph: Clownface { _context: [ [Context] ] },
  _unionGraph: Clownface { _context: [ [Context] ] },
  types: default_1 { __resource: [Circular] },
  __initializeProperties: [Function: f] { called: true, value: true }
}
```
I don't see a `name`, `delimiter` or `quote` in there. Is this normal?

**Comment by tpluscode** [comment]
Unfortunately yes, because the object is a JS Proxy.

Try `command._selfGraph.dataset.toString()`

**Comment by martinmaillard** [comment]
```
_:b71 <http://schema.org/name> "Stammdaten_tierart.csv" .
_:b71 <https://rdf-cube-curation.described.at/csvDelimiter> ";" .
_:b71 <https://rdf-cube-curation.described.at/csvQuote> "\"" .
```

**Comment by martinmaillard** [comment]
Now it's working if I provide `@id`

**Comment by martinmaillard** [comment]
One thing I find weird is that if you change the `csvDelimiter`, it doesn't seem to alter the way the sample data is parsed. I find this feature quite confusing overall.

**Comment by tpluscode** [comment]
You're right. The sample data is not being saved in the same csv dialect as the input but always using semicolon and double quotes.

Might have to revise that...

**Pull request merged**